soroban-env-host/src/host/data_helper.rs

```diff
@@ -569,4 +570,28 @@ impl Host {
         }
         Ok(())
     }
+
+    // Checks whether the given contract has a special 'dummy' executable
+    // that marks contracts created with `register_test_contract`.
+    // "testutils" are not covered by budget metering.
+    pub(crate) fn is_test_contract_executable(
+        &self,
+        contract_id: &Hash,
+    ) -> Result<bool, HostError> {
+        let key = self.contract_instance_ledger_key(contract_id)?;
+        let instance = self.retrieve_contract_instance_from_storage(&key)?;
+        let test_contract_executable = ContractExecutable::Wasm(
+            crypto::sha256_hash_from_bytes(&[], self)?
```

Comment:
❗ Am I understanding correctly that the marker for whether a contract is a test contract the hash of empty bytes? It'd be easier to eyeball it in test data if the value was simply `[0u8; 32]`. Is there a reason it needs to be an actual hash of a known preimage?

Reply:
💡 We could make it a real enum value in the XDR, i.e. `ContractExecutable::External`? That feels like it'd be less of a hack.

```c
enum ContractExecutableType
{,
    CONTRACT_EXECUTABLE_EXTERNAL = -1,
    CONTRACT_EXECUTABLE_WASM = 0,
    CONTRACT_EXECUTABLE_TOKEN = 1
};

union ContractExecutable switch (ContractExecutableType type)
{
case CONTRACT_EXECUTABLE_EXTERNAL:
    void;
case CONTRACT_EXECUTABLE_WASM:
    Hash wasm_hash;
case CONTRACT_EXECUTABLE_TOKEN:
    void;
};
```

There's no op or host function for uploading none, so it'd be not seen in pubnet, but the type system would echo a reality where envs can support arbitrary other types of contract executables that aren't stored in the ledger.

Reply:
What happens if someone uploads an empty contract? Will it think it's a test contract? Given the code has to handle this case, why do we even need a marker? Can't the presence of a registered test contract simply override whatever is on ledger? And then if a contract is updated, we can delete the registered test contract?